### PR TITLE
Adding game state

### PR DESCRIPTION
**Pull request by akhesaCaro**
Thanks to checking my branch :)

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -5,20 +5,39 @@ import  GameBoard
...
 
--- | Window Backgrohjund Color
+-- | Window background Color
 background :: Color
 background = black
 
 -- | Update the game by moving the ball and bouncing off walls.
 update :: Float   -- ^ The number of seconds since last update
-       -> Game    -- ^ The intial game state
+       -> Game    -- ^ The initial game state
```
Comment (gdeest) [inline app/Main.hs:20]: "Initial" is a bit misleading. I'd say "Current game state".

File: app/Main.hs [hunk 2]
```diff
@@ -5,20 +5,39 @@ import  GameBoard
...
+-- Game playing
+update seconds game @ Game { gameState = Playing }
+      = isGameOver
+      . movePaddle
+      . bricksBounce seconds
+      . moveBall seconds
+      . collisionBounce seconds
+      . computeDot $ game
+
+-- Game paused
```
Comment (gdeest) [inline app/Main.hs:32]: Why not just:

```haskell
update _ game = game
```

to handle the three last cases at once ?

File: src/Maths.hs [hunk 3]
```diff
@@ -29,7 +29,7 @@ cramer ((v1x, v1y), (v2x, v2y)) (rx, ry) =
       where
         det = determinant ((v1x, v1y), (v2x, v2y))
 
--- | return the intersection point between 2 vectors, if exists
+-- | return, if exist, the intersection point between 2 vectors construct each one with a point and a vector
```
Comment (gdeest) [inline src/Maths.hs:32]: Return, if it exists, the intersection point between 2 *segments* defined by a starting point and a direction vector.

Comment (gdeest) [reply src/Maths.hs:32]: Also: line too long.

File: src/Physics.hs [hunk 4]
```diff
@@ -66,7 +65,7 @@ collisionBounce s game = game { ballVel = (nsX / s , nsY / s) }
         (nsX, nsY)  = rectanglesDotCollision ballD (ballVX * s, ballVY * s) rectangles
         rectangles = [ (paddleLoc $ paddle game, paddleWidth, paddleHeight)
                      , (wallUpPos, gameWidth, wallWidth)
-                     , (wallDownPos, gameWidth, wallWidth)
+                  --   , (wallDownPos, gameWidth, wallWidth)
```
Comment (gdeest) [inline src/Physics.hs:68]: Dead code.

File: src/Physics.hs [hunk 5]
```diff
@@ -126,33 +125,12 @@ movePaddle game
...
-          (vx, vy) = paddleVel $ paddle game
-          dotLoc :: Position
-          dotLoc = (x - paddleWidth / 2, y)
-          dotSpeed :: Speed
-          dotSpeed = (vx * seconds, vy * seconds)
-          rectangle :: Rectangle
-          rectangle = ((x, y), paddleWidth, paddleHeight)
-          newSpeed = traceShowId $ rectangleDotCollision dotLoc dotSpeed rectangle
+-- | Verify if the game is over (ball outside the game)
+isGameOver :: Game  -- ^ Intial game state
```
Comment (gdeest) [inline src/Physics.hs:130]: Initial.

File: src/Physics.hs [hunk 6]
```diff
@@ -126,33 +125,12 @@ movePaddle game
...
+           -> Game  -- ^ Game updated
```
Comment (gdeest) [inline src/Physics.hs:131]: Updated game.

File: src/Rendering.hs [hunk 7]
```diff
@@ -19,8 +19,9 @@ renderDot c (x, y) radius = translate x y $ color c $  circleSolid radius
 renderStateText :: Color  -- ^ Text color
             -> String     -- ^ Text
             -> Position   -- ^ Text position
+            -> (Float, Float) -- ^ Text scale
```
Comment (gdeest) [inline src/Rendering.hs:22]: Scaling factors along X and Y dimensions.

File: src/Rendering.hs [hunk 8]
```diff
@@ -56,10 +57,27 @@ renderPaddle c w h (x, y) = translate x y $ color c $ rectangleSolid w h
 -- | render the game
 renderGame :: Game      -- ^ The game state to render
            -> Picture   -- ^ A picture of this game state
-renderGame game = pictures
+
+-- MainMenu state
+renderGame game @ Game { gameState = MainMenu } = pictures
+      [ renderStateText orange "Haskell" (-120, 100) (0.5, 0.5)
+      , renderStateText orange "Breakout" (-150, 0) (0.5, 0.5)
+      , renderStateText orange "Press ENTER to continu" (-200, -100) (0.25, 0.25)
```
Comment (gdeest) [inline src/Rendering.hs:65]: continu -> continue.

File: src/Rendering.hs [hunk 9]
```diff
@@ -56,10 +57,27 @@ renderPaddle c w h (x, y) = translate x y $ color c $ rectangleSolid w h
...
+-- Paused state
+renderGame game @ Game { gameState = Paused } =
+      renderStateText orange "PAUSED" (-120, 0) (0.5, 0.5)
+
+-- Playing state
+renderGame game @ Game { gameState = Playing } = pictures
       [ renderBall (dark red) 10 (ballLoc game)
       , renderWall wallColor gameWidth wallWidth wallUpPos
-      , renderWall wallColor gameWidth wallWidth wallDownPos
+--      , renderWall wallColor gameWidth wallWidth wallDownPos
```
Comment (gdeest) [inline src/Rendering.hs:80]: Dead code.

**Review by gdeest — APPROVED** [review]

**Comment by akhesaCaro** [comment]
Comment treated in the "change_ball_velocity" branch

**Pull request merged**